presto-main/src/main/java/com/facebook/presto/operator/ArrayBlockEncodingBuffers.java

```diff
@@ -0,0 +1,331 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.block.ColumnarArray;
+import com.facebook.presto.spi.block.DictionaryBlock;
+import com.facebook.presto.spi.block.RunLengthEncodedBlock;
+import io.airlift.slice.ByteArrays;
+import io.airlift.slice.SliceOutput;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+
+import static com.facebook.presto.operator.ByteArrayUtils.writeLengthPrefixedString;
+import static com.google.common.base.Verify.verify;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+import static sun.misc.Unsafe.ARRAY_INT_INDEX_SCALE;
+
+public class ArrayBlockEncodingBuffers
+        extends BlockEncodingBuffers
+{
+    private static final String NAME = "ARRAY";
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(ArrayBlockEncodingBuffers.class).instanceSize();
+
+    // The buffer for the offsets for all incoming blocks so far
+    private byte[] offsetsBuffer;
+
+    // The address that the next offset value will be written to.
+    private int offsetsBufferIndex;
+
+    // This array holds the condensed offsets for each position for the incoming block. It starts from the offset of row 1 instead of 0.
```

Comment:
Is there any reason for not having offsets start with zero? I think making offsets start with zero and have positions+1 entries will simply logic in many places. This will also make it similar to how offsets are handled elsewhere in the code base.